dags/process_elife_zip_dag.py

```diff
@@ -0,0 +1,107 @@
+"""
+DAG processes a zip file stored in an AWS S3 bucket and prepares the extracted xml
+file for Libero content API and sends to the content store via a PUT request.
+"""
+import logging
+from datetime import timedelta
+from io import BytesIO
+from tempfile import TemporaryFile
+from zipfile import ZipFile
+
+from airflow import DAG, configuration
+from airflow.operators import python_operator
+from airflow.utils import timezone
+from lxml import etree
+
+from aws import get_aws_connection, list_bucket_keys_iter
+
+SOURCE_BUCKET = configuration.conf.get('libero', 'source_bucket')
+DESTINATION_BUCKET = configuration.conf.get('libero', 'destination_bucket')
+
+logger = logging.getLogger(__name__)
+
+default_args = {
+    'owner': 'libero',
+    'depends_on_past': False,
+    'start_date': timezone.utcnow(),
+    'email': ['libero@example.com'],
+    'email_on_failure': False,
+    'email_on_retry': False,
+    'retries': 3,
+    'retry_delay': timedelta(seconds=5)
+}
+
+
+def extract_zipped_files_to_bucket(**context):
+    # get file name passed from trigger
+    dag_run = context['dag_run']
+    conf = dag_run.conf or {}
+    file_name = conf.get('file')
+    logger.debug('FILE PASSED= %s', file_name)
+    if not file_name:
+        raise ValueError('%s triggered without a file name passed to conf' % dag_run.dag_id)
+
+    # temporary files are securely stored on disk and automatically deleted when closed
+    with TemporaryFile() as temp_file:
+        s3 = get_aws_connection('s3')
+        # store downloaded file in temp file
+        s3.download_fileobj(SOURCE_BUCKET, file_name, temp_file)
+        logger.debug('ZIPPED FILES= %s', ZipFile(temp_file).namelist())
+
+        folder_name = file_name.replace('.zip', '')
+        xml_path = None
+        for zipped_file_path in ZipFile(temp_file).namelist():
+            # skip mac os archive files
+            if '__MACOSX' not in zipped_file_path:
```

Comment:
uh, should be extracted in a list of names to skip as hidden directories? I can think of similar examples from Windows or Linux

Reply:
Yes, I wasn't sure how far to go with this. Either have a list of patterns to skip or do not skip anything. What do you think?

Reply:
I think existing projects take a whitelist approach and have a list of patterns they look for (XML, images, and so on) ignoring all the rest. However, that's more for backward compatibility in case new types of files are added (e.g. videos or supplementary files like a dataset in .csv format); if a content processor sent .zip archives that contained empty or unused folders I would send them back as incorrect input.

Reply:
What would you say is the requirement here?

Reply:
Path of least resistance: take the files you need and process them; if additional, unnecessary files are making the code fail, let it fail.

Reply:
added a list of accepted file extensions.